src/gt4py/next/ffront/foast_to_itir.py

```diff
@@ -318,12 +318,22 @@ def visit_Call(self, node: foast.Call, **kwargs) -> itir.Expr:
     def _visit_astype(self, node: foast.Call, **kwargs) -> itir.FunCall:
```

Comment:
This I am not sure about, but it might be that the return type could be `itir.Expr` in some cases, which would supersed `itir.FunCall`. Could you please check?

Reply:
That's a good point. When developing the` _visit_astype` method, `mypy` who checks the return types in `pre-commit` complained that `itir.Expr` was returned when `itir.FunCall` is expected. I wanted to use the `ir_makers.promote_to_lifted_stencil` function but it returns an `itir.Expr`, so I finally used its contents without calling it and the tests are now green. So no `itir.Expr` should be returned to answer your question but doesn't the root of this problem lie in the fact that the return type of `im.promote_to_lifted_stencil` which also is called in the `_map` function should rather be a itir.`FunCall `as it calls a lift rather than `itir.Expr`?

Reply:
this is more of a question for @tehrengruber

Reply:
As a general guideline: A type annotation should not affect the way you write code, but merely help in finding accidental errors. Occasionally (or given the terrible type system in python frequently) the type checker needs some help and a cast or assert is needed, but that's about it. With respect to the actual problem: The annotation in `promote_to_lifted_stencil` is too lose. In fact it only ever returns an `itir.FunCall` (or callable returning one to be precise) , which is why replacing the call to `promote_to_lifted_stencil` with its implementation made mypy happy. If you tighten the annotation in `promote_to_lifted_stencil` to `itir.FunCall` it should work.